examples/model_checking.py

```diff
@@ -35,115 +36,205 @@ def __init__(self, variables, init, trans):
-                v_i = v.substitute(subs_i)
-                v_j = v.substitute(subs_j)
-                state.append(Not(EqualsOrIff(v_i, v_j)))
-            res.append(Or(state))
-    return And(res)
-
-
-def get_k_hypothesis(system, prop, k):
-    """Hypothesis for k-induction: each state up to k-1 fulfills the property"""
-    res = []
-    for i in xrange(k):
-        subs_i = get_subs(system, i)
-        res.append(prop.substitute(subs_i))
-    return And(res)
-
-
-def get_bmc(system, prop, k):
-    """Returns the BMC encoding at step k"""
-    init_0 = system.init.substitute(get_subs(system, 0))
-    prop_k = prop.substitute(get_subs(system, k))
-    return And(get_unrolling(system, k), init_0, Not(prop_k))
-
-def get_k_induction(system, prop, k):
-    """Returns the K-Induction encoding at step K"""
-    subs_k = get_subs(system, k)
-    prop_k = prop.substitute(subs_k)
-    return And(get_unrolling(system, k),
-               get_k_hypothesis(system, prop, k),
-               get_simple_path(system, k),
-               Not(prop_k))
-
-def check_property(system, prop):
-    """Interleaves BMC and K-Ind to verify the property."""
-    print("Checking property %s..." % prop)
-    for b in xrange(100):
-        f = get_bmc(system, prop, b)
-        print("   [BMC]    Checking bound %d..." % (b+1))
-        if is_sat(f):
-            print("--> Bug found at step %d" % (b+1))
-            return
-
-        f = get_k_induction(system, prop, b)
-        print("   [K-IND]  Checking bound %d..." % (b+1))
-        if is_unsat(f):
-            print("--> The system is safe!")
-            return
-
-
-def main():
-    # Example Transition System (SMV-like syntax)
+class PDR(object):
+    def __init__(self, system):
+        self.system = system
+        self.R = [system.init]
+        self.solver = Solver()
+        
+    def check_property(self, prop):
+        """Property Directed Reachability approach without optimizations."""
+        print("Checking property %s..." % prop)
+
+        while True:
+            cube = self.get_bad_state(prop)
+            if cube is not None:
+                if self.recursive_block(cube):
+                    print("--> Bug found at step %d" % (len(self.R)))
+                    return
+            else:
+                if self.inductive(prop):
+                    print("--> The system is safe!")
+                    return
+                print("   [PDR]    Adding frame %d..." % (len(self.R)))
+                self.R.append(TRUE())
+
+    def get_bad_state(self, prop):
+        """Extracts a reachable state that intersects the negation 
+        of the property and the last current frame"""
+        return self.solve(And(self.R[-1], Not(prop)))
+
+    def solve(self, formula):
+        """Provides a satisfiable assignment to the state variables that are consistent with the input formula"""
+        self.solver.reset_assertions()
+        self.solver.add_assertion(formula)
+        if self.solver.solve():
```

Comment:
You should be able to simplify the above by doing ```self.solver.solve(formula)```. This will solve under the assumption of formula.